File: include/apl/errors/ImplementationError.java
package APL.errors;

import APL.types.*;

public class ImplementationError extends APLError {
  public ImplementationError(std::string s) {
    super(s);
  }
  public ImplementationError(std::string s, Tokenable fun) {
    super(s, fun);
  }
  public ImplementationError(std::string s, Callable fun, Tokenable cause) {
    super(s, fun, cause);
  }
  public ImplementationError(Throwable t) {
    super(t.getMessage());
    initCause(t);
  }
}

File: APL/errors/ImplementationError.cpp
#include <APL/ImplementationError.h>

namespace APL::errors
{
    // using namespace APL::types;

    ImplementationError::ImplementationError(std::string const& s) : APLError(s)
    {
    }

    ImplementationError::ImplementationError(std::string const& s, std::shared_ptr<APL::types::Tokenable> fun) : APLError(s, fun)
    {
    }

    ImplementationError::ImplementationError(std::string const& s, std::shared_ptr<APL::types::Callable> fun, std::shared_ptr<APL::types::Tokenable> cause) : APLError(s, fun, cause)
    {
    }

    ImplementationError::ImplementationError(std::runtime_error t) : APLError(t.what())
    {
      initCause(t);
    }
}
